README.md

```diff
@@ -67,15 +68,67 @@ argument. Some useful kernel aguments:
 
 * `init=PROGRAM` - Specifies the userspace program for PID 1. Browse `./user`
   for currently available programs.
-* `test=TEST` - Requests the kernel to run the specified test (from `./tests`
-  directory).
+* test-related arguments described in the [following section](#test-infrastructure).
+
+Test infrastructure
+---
+
+##### User tests
+Located in `/user/utest`.
+User-space test function signature looks like this: `int test_[name](void)` 
+and should be defined in `user/utest/utest.h`.
+In order to make the test runnable one has to add one of these lines to `test/utest.c` file:
+* `UTEST_ADD_SIMPLE([name])` - test fails on assertion or non-zero return value.
+* `UTEST_ADD_SIGNAL([name], [SIGNUMBER])` - test passes when terminated with `[SIGNUMBER]`.
+* `UTEST_ADD([name], [exit status], flags)` - test passes when exited with status `[exit status]`.
+
+One also needs to add a line `CHECKRUN_TEST([name])` in `/user/utest/main.c`.
+
+##### Kernel tests 
+Located in `/tests`. 
+Test function signature looks like this:
+`[name](void)` or sometimes `[name](unsigned int)` but needs to be casted to 
+`(int (*)(void))`.
+
+Macros to register tests:
+* `KTEST_ADD(name, func, flags)`
+* `KTEST_ADD_RANDINT(name, func, flags, max)` - need to cast function pointer to
+`(int (*)(void))`
+
+Where `name` is test name, `func` is pointer to test function, 
+flags as mentioned bellow, and `max` is maximum random argument fed to the test.
+
+##### Flags
+* `KTEST_FLAG_NORETURN` - signifies that a test does not return.
+* `KTEST_FLAG_DIRTY` - signifies that a test irreversibly breaks internal kernel state, and any
+   further test done without restarting the kernel will be inconclusive.
+* `KTEST_FLAG_USERMODE` - indicates that a test enters usermode.
+* `KTEST_FLAG_BROKEN` - excludes the test from being run in auto mode. This flag is only useful for
+   temporarily marking some tests while debugging the testing framework.
+* `KTEST_FLAG_RANDINT` - marks that the test wishes to receive a random integer as an argument.
+
+##### Running tests:
+* `./run_tests.py` - runs 5 seeds of all tests.
+* `./run_tests.py --infinite` - infinitely runs all tests.
+* `./run_tests.py --non-interactive` - do not run gdb session if tests fail.
```

Comment:
Maybe it's worth saying, that --non-interactive shows the same gdb output, but without possibility to interact. "do not run gdb session" sounds like it doesn't shows gdb output at all.